### PR TITLE
Improve/fix authorization error handling

### PR DESCRIPTION
**Pull request by galfert**
This is still WIP, because I'm still working on the Widget-side of things and I don't know yet if there is anything else missing. Also, there are no tests yet.

### CODE REVIEW

File: src/authorize.js [hunk 1]
```diff
@@ -166,7 +166,11 @@ var util = require('./util');
       var authParamsUsed = false;
       if (params) {
         if (params.error) {
-          throw "Authorization server errored: " + params.error;
+          if (params.error === 'access_denied') {
+            throw new Authorize.Unauthorized('User denied access.');
```
Comment (skddc) [inline src/authorize.js]: How about "Authorization failed: access denied".

File: src/authorize.js [hunk 2]
```diff
@@ -166,7 +166,11 @@ var util = require('./util');
...
+          } else {
+            throw "Authorization server errored: " + params.error;
```
Comment (skddc) [inline src/authorize.js]: I think "Authorization failed: [error]" is a bit more concise and human-friendly.

**Comment by galfert** [comment]
I added a fix to not remove any bearer token and/or configuration settings when getting an authorization error. Otherwise, the error wouldn't be shown to users anymore after reloading the app.

**Comment by skddc** [comment]
So is it still WIP?

**Comment by galfert** [comment]
It should all work now, I just didn't look into writing any tests for the changes yet.

**Comment by skddc** [comment]
Ok, cool.

**Comment by galfert** [comment]
I added an optional `code` property to the `Unauthorized` error object.

When a user denies the authorization, the code `access_denied` is added, so the widget can easier detect that particular error.

I tested this in Chrome, Firefox, Safari and Opera. They all can see the additional property on the error. According to [this article](https://medium.com/@xjamundx/custom-javascript-errors-in-es6-aa891b173f87) it shouldn't be a problem to add custom properties to an error object, it just can't have custom methods.

**Comment by skddc** [comment]
That article is about node.js. The potential problems are with e.g. Internet Explorer, Opera Mobile, Android browser, and such.

**Comment by galfert** [comment]
Update: It also works on Edge. But on IE 11 I already get an error during discovery. Still trying to debug that. Looks like it's a problem in Webfinger.js.

**Comment by galfert** [comment]
I fixed the Webfinger.js issue in IE 11 with silverbucket/webfinger.js#27.

After that all worked fine, including the additional error property. I just had to bring my own Promises :)

**Comment by skddc** [comment]
Yay! Good job. 🎉

**Comment by galfert** [comment]
Tested Opera Mobile and Android Browser on Android 4.1.2 and no problems there either.

**Comment by skddc** [comment]
Great! I think we can all agree that's enough then, right?

**Comment by galfert** [comment]
I think so, yes.

**Comment by galfert** [comment]
I added some tests as well, so I think this is ready to merge now.

**Comment by skddc** [comment]
Fantastic!

**Pull request merged**